File: 622-design-circular-queue/622-design-circular-queue.cpp
class MyCircularQueue {
public:
    int front = 0;
    int rear = 0;
    int len;
    vector<int> queue;
    MyCircularQueue(int k) {
        for (int i=0; i<k; i++) {
            queue.push_back(-1);
        }
        this->len = k;
    }
    
    bool enQueue(int value) {
        if (this->isFull()) {
            return false;
        }
        if (!(this->queue[this->rear] == -1)) {
            this->rear = (this->rear + this->len - 1) % this->len;
        }
        this->queue[this->rear] = value;
        return true;
    }
    
    bool deQueue() {
        if (this->isEmpty()) {
            return false;
        }
        this->queue[this->front] = -1;
        if (this->front != this->rear) {
            this->front = (this->front + this->len - 1) % this->len;
        }
        return true;
    }
    
    int Front() {
        if (!(this->queue[this->front] == -1)) {
            return this->queue[this->front];
        }
        return -1;
    }
    
    int Rear() {
        if (!(this->queue[this->rear] == -1)) {
            return this->queue[this->rear];
        }
        return -1;
    }
    
    bool isEmpty() {
        return ((this->front == this->rear) & (this->queue[this->front] == -1));
    }
    
    bool isFull() {
        return (((this->front == (this->len - 1)) & (this->rear == 0)) || (this->front == (this->rear - 1))) & !(this->queue[this->front] == -1) & !(this->queue[this->rear] == -1);
    }
};

/**
 * Your MyCircularQueue object will be instantiated and called as such:
 * MyCircularQueue* obj = new MyCircularQueue(k);
 * bool param_1 = obj->enQueue(value);
 * bool param_2 = obj->deQueue();
 * int param_3 = obj->Front();
 * int param_4 = obj->Rear();
 * bool param_5 = obj->isEmpty();
 * bool param_6 = obj->isFull();
 */

File: 622-design-circular-queue/622-design-circular-queue.py
class MyCircularQueue:

    def __init__(self, k: int):
        self.queue = [None for _ in range(k)]
        self.front = 0
        self.rear = 0
        self.len = k

    def enQueue(self, value: int) -> bool:  # add to the rear
        if self.isFull():
            return False
        if not(self.queue[self.rear] is None):
            self.rear = (self.rear - 1) % self.len
        self.queue[self.rear] = value
        return True

    def deQueue(self) -> bool:  # remove from the front
        if self.isEmpty():
            return False
        self.queue[self.front] = None
        if self.front != self.rear:
            self.front = (self.front - 1) % self.len
        return True

    def Front(self) -> int:
        if not(self.queue[self.front] is None):
            return self.queue[self.front]
        return -1

    def Rear(self) -> int:
        if not(self.queue[self.rear] is None):
            return self.queue[self.rear]
        return -1

    def isEmpty(self) -> bool:
        return (self.front == self.rear) and (self.queue[self.front] is None)

    def isFull(self) -> bool:
        return (((self.front == (self.len - 1)) and (self.rear == 0)) or (self.front == (self.rear - 1))) and not(self.queue[self.front] is None) and not(self.queue[self.rear] is None)


# Your MyCircularQueue object will be instantiated and called as such:
# obj = MyCircularQueue(k)
# param_1 = obj.enQueue(value)
# param_2 = obj.deQueue()
# param_3 = obj.Front()
# param_4 = obj.Rear()
# param_5 = obj.isEmpty()
# param_6 = obj.isFull()